scmrepo/git/backend/dulwich/__init__.py

```diff
@@ -72,19 +72,22 @@ def mode(self) -> int:
         return self._mode
 
     def scandir(self) -> Iterable["DulwichObject"]:
-        tree = self.repo[self.sha]
+        tree = self.repo[self._sha]
         for entry in tree.iteritems():  # noqa: B301
             yield DulwichObject(
                 self.repo, entry.path.decode(), entry.mode, entry.sha
             )
 
     @cached_property
     def size(self) -> int:  # pylint: disable=invalid-overridden-method
-        return len(self.repo[self.sha].as_raw_string())
+        try:
+            return self.repo[self._sha].raw_length()
+        except KeyError:
+            return 0
 
-    @property
-    def sha(self) -> str:
-        return self._sha
+    @cached_property
```

Comment:
Do we really need a `cached_property` here?

Reply:
@efiop, could you please share some benchmarks?

Reply:
@skshetry No benchmarks there. Dulwich is not even used usually here (pygit2 is). Just from a correctness perspective.